notebooks/config.ipynb

```diff
@@ -0,0 +1,1046 @@
+    "<!--cell 32-->"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 17,
+   "id": "f4538965-d3b0-4c0c-a878-b6852f8d8ab0",
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "&memo\n",
+      "    sender_id = '{{ id }}'\n",
+      "    message = 'Salutations, Mars!'\n",
+      "    sent = .true.\n",
+      "    greeting = 'Salutations'\n",
+      "    recipient = 'Mars'\n",
+      "/\n"
+     ]
+    }
+   ],
+   "source": [
+    "%%bash\n",
+    "cat tmp/updated-config.nml"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "a20024ac-0f33-4000-942d-29c99dc0502e",
+   "metadata": {},
+   "source": [
+    "### Using the `key_path` parameter\n",
+    "\n",
+    "Consider the following config file, where the desired keys and values may not be at the top level.\n",
+    "<!--cell 34-->"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 18,
+   "id": "7ce0e917-f0d0-4302-9c8c-b136ffc5410a",
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "keys:\n",
+      "  to:\n",
+      "    config:\n",
+      "      message: \"{{ greeting }}, {{ recipient }}!\"\n"
+     ]
+    }
+   ],
+   "source": [
+    "%%bash\n",
+    "cat fixtures/config/keys-config.yaml"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "633778ca-391c-4ba4-b3b4-b3f599af5d41",
+   "metadata": {},
+   "source": [
+    "Using `key_path` allows only the portion of the config contained within the given list of keys to be written to a file or, in this case, to `stdout`. Note that the key-value pairs from the update config are used to render values and appear in the returned config dictionary, but they don't appear in the config written to `stdout`.\n",
+    "<!--cell 36-->"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 19,
+   "id": "00150efa-848c-44eb-ac0c-dab3845546b8",
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "message: Good morning, Venus!\n"
+     ]
+    },
+    {
+     "data": {
+      "text/plain": [
+       "{'keys': {'to': {'config': {'message': 'Good morning, Venus!'}}},\n",
+       " 'greeting': 'Good morning',\n",
+       " 'recipient': 'Venus'}"
+      ]
+     },
+     "execution_count": 19,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "config.realize(\n",
+    "    input_config=\"fixtures/config/keys-config.yaml\",\n",
+    "    update_config={\"greeting\": \"Good morning\", \"recipient\": \"Venus\"},\n",
+    "    output_file=None,\n",
+    "    output_format='yaml',\n",
+    "    key_path=['keys', 'to', 'config']\n",
+    ")"
```

Comment:
I agree that having the two outputs together there maybe isn't the best. What would you think of splitting that demo into two code cells? I can assign the dictionary returned by `realize()` to a variable that I can print in a follow-up code cell so that I don't clutter the `stdout` output. Or do you think the `dict` isn't worth showing there?

Reply:
I don't think the `dict` is worth showing in this section: I think the `config.realize_to_dict()` section shows that off well. This section is about producing _external_ output, either in a file or on `stdout`, and the fact that `config.realize()` happens to return a `dict` is incidental in this context. (It's not incidental in `uwtools`, which takes advantage of that fact to define `realize_to_dict()` as a specially crafted call to `realize()`.)